### PR TITLE
Refactor api coin and token router

### PR DESCRIPTION
**Pull request by boqiu**
1. Decouple `API Coin` and `TokenRouter`.
2. Leverage `ERC4626` for `API Coin` to supports deposit/withdraw based on a specified token, e.g. USDT.
3. Supports to deposit/withdraw `CFX` with swap based exchange.

- - -
This change is [Reviewable](https://reviewable.io/reviews/conflux-chain/web3pay-contract/39)

### CODE REVIEW

File: contracts/v2/SwapExchange.sol [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "./AppCoin.sol";
+
+/** The interface of swapping contract (SwappiRouter on Conflux eSpace). */
+interface ISwap {
+    // ETH => Token
+    function swapETHForExactTokens(uint amountOut, address[] calldata path, address to, uint deadline)
```
Comment (Tigo9527) [inline contracts/v2/SwapExchange.sol:11]: Missing return value?

File: contracts/v2/SwapExchange.sol [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        amount = appCoin.previewMint(amount);
+
+        uint256 balanceBefore = address(this).balance - msg.value;
+
+        address[] memory path = new address[](2);
+        path[0] = swap.WETH();
+        path[1] = appCoin.asset();
+
+        // swap ETH for tokens
+        swap.swapETHForExactTokens{value: msg.value}(amount, path, receiver, block.timestamp);
```
Comment (Tigo9527) [inline contracts/v2/SwapExchange.sol:88]: Receiver should be address(this)

File: contracts/v2/SwapExchange.sol [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+        // approve and deposit for receiver
+        SafeERC20.safeApprove(IERC20(appCoin.asset()), address(appCoin), amount);
+        appCoin.deposit(amount, receiver);
+
+        // refund dust ETH if any
+        uint256 dust = address(this).balance - balanceBefore;
```
Comment (Tigo9527) [inline contracts/v2/SwapExchange.sol:95]: Check consumed amount of eth in return value from swapETHForExactTokens

File: contracts/v2/SwapExchange.sol [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    function previewWithdrawETH(uint256 amount) public view onlyInitialized returns (uint256) {
+        amount = appCoin.previewRedeem(amount);
+
+        address[] memory path = new address[](2);
+        path[0] = appCoin.asset();
+        path[1] = swap.WETH();
+
+        uint[] memory amounts = swap.getAmountsOut(amount, path);
+
+        return amounts[0];
```
Comment (Tigo9527) [inline contracts/v2/SwapExchange.sol:116]: Should return last element

**Review by Tigo9527 — APPROVED** [review]

File: contracts/v2/SwapExchange.sol [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+/**
+ * @dev SwapExchage is used to deposit/withdraw App Coins based on native tokens.
+ */
+contract SwapExchage {
+    AppCoin private appCoin;
+    ISwap private swap;
+
+    modifier onlyInitialized() {
```
[the remark this answers is not in the document]
Comment (boqiu) [reply contracts/v2/SwapExchange.sol]: Removed.

File: contracts/v2/SwapExchange.sol (hunk 1 above)
Comment (boqiu) [reply contracts/v2/SwapExchange.sol:11]: See the next line below.

File: contracts/v2/SwapExchange.sol (hunk 2 above)
Comment (boqiu) [reply contracts/v2/SwapExchange.sol]: Done.

File: contracts/v2/SwapExchange.sol (hunk 3 above)
Comment (boqiu) [reply contracts/v2/SwapExchange.sol:93]: Relying on contract balance is more error tolerable, we can have more discussion offline.

File: contracts/v2/SwapExchange.sol (hunk 4 above)
Comment (boqiu) [reply contracts/v2/SwapExchange.sol]: Fixed.

File: contracts/v2/SwapExchange.sol (hunk 3 above)
Comment (boqiu) [reply contracts/v2/SwapExchange.sol:94]: In addition, add `ReentrancyGuard` to prevent malicious callback from `receiver` contract.

**Review by Tigo9527 — APPROVED** [review]

**Pull request merged**